File: components/RecurringTasks/streaks/patterns/weekly.ts
import { startOfDay, endOfDay } from 'date-fns';
import { StreakResult, WeeklyRecurrencePattern, DateRange, RecurrenceTypes, WeekDay } from '../types';
import { WEEKDAYS } from '../helpers/constants';
import { isWeeklyPattern } from './patternMatchers';

interface WeeklyTarget {
  date: Date;
  allowedRange: DateRange;
  weekday: WeekDay;
}

// Add weekday name mapping
const WEEKDAY_NAMES: Record<string, string> = {
  'sun': 'SUNDAY',
  'sunday': 'SUNDAY',
  'mon': 'MONDAY',
  'monday': 'MONDAY',
  'tue': 'TUESDAY',
  'tues': 'TUESDAY',
  'tuesday': 'TUESDAY',
  'wed': 'WEDNESDAY',
  'wednesday': 'WEDNESDAY',
  'thu': 'THURSDAY',
  'thur': 'THURSDAY',
  'thurs': 'THURSDAY',
  'thursday': 'THURSDAY',
  'fri': 'FRIDAY',
  'friday': 'FRIDAY',
  'sat': 'SATURDAY',
  'saturday': 'SATURDAY'
};

export function calculateWeeklyStreak(
  pattern: WeeklyRecurrencePattern,
  completions: Date[],
  range: DateRange
): StreakResult {
  if (pattern.type !== RecurrenceTypes.WEEKLY) {
    throw new Error('Invalid pattern type for weekly streak calculation');
  }

  // Generate target dates based on pattern
  const targetDates = generateWeeklyTargets(pattern, range);
  if (!targetDates.length) {
    return { currentStreak: 0, longestStreak: 0 };
  }

  // Sort completions from newest to oldest
  const sortedCompletions = [...completions]
    .map(date => startOfDay(date))
    .sort((a, b) => b.getTime() - a.getTime());
  
  const now = startOfDay(new Date());
  
  let currentStreak = 0;
  let longestStreak = 0;
  let tempStreak = 0;
  let activeStreak = true;

  // Process targets from newest to oldest
  for (const target of targetDates) {
    const targetDay = startOfDay(target.date);
    
    // Skip future targets
    if (targetDay > now) {
      continue;
    }

    // Check if this target was completed
    const isCompleted = sortedCompletions.some(completion => 
      completion.getTime() === targetDay.getTime()
    );

    if (isCompleted) {
      tempStreak++;
      if (activeStreak) {
        currentStreak = tempStreak;
      }
      longestStreak = Math.max(longestStreak, tempStreak);
    } else {
      // For the most recent target, don't break streak if we're still within the day
      if (targetDay.getTime() === now.getTime()) {
        if (tempStreak > 0) {
          currentStreak = tempStreak;
          longestStreak = Math.max(longestStreak, tempStreak);
        }
      } else {
        activeStreak = false;
        tempStreak = 0;
      }
    }
  }

  return { currentStreak, longestStreak };
}

function generateWeeklyTargets(pattern: WeeklyRecurrencePattern, range: DateRange): WeeklyTarget[] {
  const targets: WeeklyTarget[] = [];
  const interval = pattern.interval || 1;
  const weekdays = pattern.weekdays;
  let currentDate = startOfDay(range.end);
  const rangeStart = startOfDay(range.start);

  // For bi-weekly patterns, we need to find a reference point
  // This will be the most recent occurrence of the pattern
  let referenceDate = new Date(currentDate);
  while (referenceDate.getDay() !== weekdays[0]) {
    referenceDate.setDate(referenceDate.getDate() - 1);
  }

  while (currentDate >= rangeStart) {
    // For bi-weekly patterns, check if this week is in sequence
    const weekDiff = Math.floor((referenceDate.getTime() - currentDate.getTime()) / (7 * 24 * 60 * 60 * 1000));
    const isTargetWeek = weekDiff % interval === 0;

    if (isTargetWeek) {
      for (const weekday of weekdays) {
        const targetDate = new Date(currentDate);
        // Adjust to the target weekday
        const diff = weekday - targetDate.getDay();
        targetDate.setDate(targetDate.getDate() + diff);
        
        // Skip if target date falls outside our range
        if (targetDate > range.end || targetDate < rangeStart) {
          continue;
        }

        targets.push({
          date: targetDate,
          allowedRange: {
            start: startOfDay(targetDate),
            end: endOfDay(targetDate)
          },
          weekday
        });
      }
    }
    
    // Move to previous week
    currentDate.setDate(currentDate.getDate() - 7);
  }

  // Sort targets by date descending (newest first)
  return targets.sort((a, b) => b.date.getTime() - a.date.getTime());
}

export function parseWeeklyPattern(pattern: string): WeeklyRecurrencePattern {
  if (!isWeeklyPattern(pattern)) {
    console.log('Pattern failed isWeeklyPattern check:', pattern);
    throw new Error('Invalid weekly pattern format');
  }

  const normalizedPattern = pattern.trim().toLowerCase();
  console.log('Attempting to parse pattern:', normalizedPattern);

  // Handle simple "every week" pattern
  if (normalizedPattern === 'every week') {
    return {
      type: RecurrenceTypes.WEEKLY,
      interval: 1,
      weekdays: [1] // Default to Monday
    };
  }

  // Handle "every X weeks" pattern
  const weekIntervalMatch = normalizedPattern.match(/^every\s+(\d+)\s+weeks?$/i);
  if (weekIntervalMatch && weekIntervalMatch[1]) {
    const interval = parseInt(weekIntervalMatch[1], 10);
    if (isNaN(interval) || interval <= 0) {
      throw new Error('Invalid week interval');
    }
    return {
      type: RecurrenceTypes.WEEKLY,
      interval,
      weekdays: [1] // Default to Monday
    };
  }

  const weekdays: WeekDay[] = [];

  // Handle bi-weekly patterns (e.g., "every other monday")
  const biWeeklyRegex = /every\s+other\s+(\w+)(?:\s+at\s+(\d{1,2})(?::(\d{2}))?\s*(am|pm)?)?/i;
  const biWeeklyMatch = normalizedPattern.match(biWeeklyRegex);

  if (biWeeklyMatch && biWeeklyMatch[1]) {
    const weekdayName = biWeeklyMatch[1].toLowerCase();
    const weekdayKey = Object.keys(WEEKDAYS).find(key => 
      key.toLowerCase() === weekdayName || 
      WEEKDAY_NAMES[weekdayName] === key
    );

    if (!weekdayKey || !(weekdayKey in WEEKDAYS)) {
      throw new Error(`Invalid weekday: ${weekdayName}`);
    }

    const weekday = WEEKDAYS[weekdayKey];
    if (typeof weekday !== 'number') {
      throw new Error(`Could not map weekday: ${weekdayKey}`);
    }

    weekdays.push(weekday as WeekDay);

    return {
      type: RecurrenceTypes.WEEKLY,
      interval: 2,
      weekdays
    };
  }

  // Handle multiple weekdays (e.g., "every monday, wednesday, friday" or "every mon, wed, fri")
  const weekdayList = normalizedPattern.replace(/^every\s+/, '').split(/,\s*|\s+and\s+/);
  for (const weekdayName of weekdayList) {
    const cleanWeekdayName = weekdayName.trim().toLowerCase();
    const weekdayKey = Object.keys(WEEKDAYS).find(key => 
      key.toLowerCase() === cleanWeekdayName || 
      WEEKDAY_NAMES[cleanWeekdayName] === key
    );

    if (!weekdayKey || !(weekdayKey in WEEKDAYS)) {
      continue; // Skip invalid weekdays
    }

    const weekday = WEEKDAYS[weekdayKey];
    if (typeof weekday !== 'number') {
      continue; // Skip invalid weekday numbers
    }

    if (!weekdays.includes(weekday as WeekDay)) {
      weekdays.push(weekday as WeekDay);
    }
  }

  if (weekdays.length > 0) {
    return {
      type: RecurrenceTypes.WEEKLY,
      interval: 1,
      weekdays: weekdays.sort((a, b) => a - b) // Sort weekdays in order
    };
  }

  throw new Error('Could not parse weekly pattern');
}

File: test/scripts/patterns/weekly.py
from datetime import datetime, timedelta
from typing import List, Optional

class WeeklyPattern:
    WEEKDAY_MAP = {
        'monday': 0, 'tuesday': 1, 'wednesday': 2, 'thursday': 3,
        'friday': 4, 'saturday': 5, 'sunday': 6,
        'mon': 0, 'tue': 1, 'wed': 2, 'thu': 3, 'fri': 4, 'sat': 5, 'sun': 6
    }

    @staticmethod
    def generate_every_other_weekday(
        start_date: datetime,
        end_date: datetime,
        weekday: str,
    ) -> List[datetime]:
        """Generate dates for every other occurrence of a specific weekday"""
        dates = []
        target_weekday = WeeklyPattern.WEEKDAY_MAP.get(weekday.lower())
        
        if target_weekday is None:
            raise ValueError(f"Invalid weekday: {weekday}")
        
        print(f"\nDebug: Generating every other {weekday}")
        print(f"Start date: {start_date.strftime('%Y-%m-%d (%A)')}")
        print(f"End date: {end_date.strftime('%Y-%m-%d (%A)')}")
        print(f"Target weekday: {target_weekday}")
        
        # Start from the beginning of the week of start_date
        current = start_date - timedelta(days=start_date.weekday())
        print(f"Adjusted to start of week: {current.strftime('%Y-%m-%d (%A)')}")
        
        # Move to the first target weekday
        while current.weekday() != target_weekday:
            current += timedelta(days=1)
        print(f"Moved to first target weekday: {current.strftime('%Y-%m-%d (%A)')}")
        
        # If we've moved past the start date, go back by two weeks
        if current < start_date:
            while current <= start_date:
                current += timedelta(weeks=2)
            current -= timedelta(weeks=2)
            print(f"Adjusted for start date: {current.strftime('%Y-%m-%d (%A)')}")
        
        # Generate dates for every other week
        print("\nGenerating dates:")
        while current <= end_date:
            if current >= start_date:
                dates.append(current)
                print(f"Added: {current.strftime('%Y-%m-%d (%A)')}")
            current += timedelta(weeks=2)
        
        return dates

    @staticmethod
    def generate_every_weekday(
        start_date: datetime,
        end_date: datetime,
        weekday: str,
        time_str: Optional[str] = None
    ) -> List[datetime]:
        """Generate dates for weekly patterns like 'every monday'"""
        dates = []
        
        # Convert weekday name to number (0 = Monday, 6 = Sunday)
        weekday_num = WeeklyPattern.WEEKDAY_MAP.get(weekday.lower())
        
        if weekday_num is None:
            raise ValueError(f"Invalid weekday: {weekday}")
        
        # Parse time if provided
        hour = 0
        minute = 0
        if time_str:
            # Handle 12-hour format with am/pm
            if 'am' in time_str.lower() or 'pm' in time_str.lower():
                time_str = time_str.lower().replace('am', '').replace('pm', '').strip()
                is_pm = 'pm' in time_str.lower()
                hour = int(time_str.split(':')[0]) % 12
                if is_pm:
                    hour += 12
                if ':' in time_str:
                    minute = int(time_str.split(':')[1])
            # Handle 24-hour format
            else:
                parts = time_str.split(':')
                hour = int(parts[0])
                if len(parts) > 1:
                    minute = int(parts[1])
        
        print("\nDebug: Generating every", weekday)
        print("Start date:", start_date.strftime("%Y-%m-%d (%A)"))
        print("End date:", end_date.strftime("%Y-%m-%d (%A)"))
        print("Target weekday:", weekday_num)
        
        # Start from the beginning of the week containing start_date
        current = start_date - timedelta(days=start_date.weekday())
        print("Adjusted to start of week:", current.strftime("%Y-%m-%d (%A)"))
        
        # Move to first target weekday
        if current.weekday() <= weekday_num:
            current += timedelta(days=weekday_num - current.weekday())
        else:
            current += timedelta(days=7 - (current.weekday() - weekday_num))
        print("Moved to first target weekday:", current.strftime("%Y-%m-%d (%A)"))
        
        # Adjust if we're before start_date
        if current < start_date:
            days_to_add = (weekday_num - start_date.weekday()) % 7
            if days_to_add == 0:
                days_to_add = 7
            current = start_date + timedelta(days=days_to_add)
        print("Adjusted for start date:", current.strftime("%Y-%m-%d (%A)"))
        
        print("\nGenerating dates:")
        while current <= end_date:
            # Set the time for this date
            target_date = current.replace(hour=hour, minute=minute)
            if start_date <= target_date <= end_date:
                print("Added:", target_date.strftime("%Y-%m-%d (%A)"))
                dates.append(target_date)
            current += timedelta(days=7)
        
        return dates

    @staticmethod
    def generate_every_n_weekday(
        start_date: datetime,
        end_date: datetime,
        weekday: str,
        interval: int
    ) -> List[datetime]:
        """Generate dates for every N weeks on a specific weekday"""
        dates = []
        target_weekday = WeeklyPattern.WEEKDAY_MAP.get(weekday.lower())
        
        if target_weekday is None:
            raise ValueError(f"Invalid weekday: {weekday}")
        
        if interval < 1:
            raise ValueError(f"Interval must be at least 1, got {interval}")
        
        print(f"\nDebug: Generating every {interval} weeks on {weekday}")
        print(f"Start date: {start_date.strftime('%Y-%m-%d (%A)')}")
        print(f"End date: {end_date.strftime('%Y-%m-%d (%A)')}")
        print(f"Target weekday: {target_weekday}")
        print(f"Interval: {interval}")
        
        # Start from the beginning of the week of start_date
        current = start_date - timedelta(days=start_date.weekday())
        print(f"Adjusted to start of week: {current.strftime('%Y-%m-%d (%A)')}")
        
        # Move to the first target weekday
        while current.weekday() != target_weekday:
            current += timedelta(days=1)
        print(f"Moved to first target weekday: {current.strftime('%Y-%m-%d (%A)')}")
        
        # If we've moved past the start date, go back by N weeks
        if current < start_date:
            while current <= start_date:
                current += timedelta(weeks=interval)
            current -= timedelta(weeks=interval)
            print(f"Adjusted for start date: {current.strftime('%Y-%m-%d (%A)')}")
        
        # Generate dates for every N weeks
        print("\nGenerating dates:")
        while current <= end_date:
            if current >= start_date:
                dates.append(current)
                print(f"Added: {current.strftime('%Y-%m-%d (%A)')}")
            current += timedelta(weeks=interval)
        
        return dates
